core/internal/filetransfer/tasks.go

```diff
@@ -14,63 +8,29 @@ const (
 )
 
 // Task is a task to upload/download a file
-type Task struct {
-	// FileKind is the category of file being uploaded or downloaded
-	FileKind RunFileKind
-
-	// Type is the type of task (upload or download)
-	Type TaskType
-
-	// Path is the local path to the file
-	Path string
-
-	// Name is the name of the file
-	Name string
-
-	// Url is the endpoint to upload to/download from
-	Url string
+type Task interface {
+	// Execute does the action (upload/download) described by the task
+	Execute(*FileTransfers) error
 
-	// Headers to send on the upload
-	Headers []string
+	// Complete executes any callbacks necessary to complete the task
+	Complete(FileTransferStats)
 
-	// Size is the number of bytes to upload
-	//
-	// If this is zero, then all bytes starting at `Offset` are uploaded; if non-zero,
-	// then that many bytes starting from `Offset` are uploaded.
-	Size int64
+	// String describes the task
+	String() string
 
-	// Offset is the beginning of the file segment to upload
-	Offset int64
-
-	// Response is the http.Response from a successful upload or download request.
-	//
-	// This is nil for failed requests, or requests that have not completed.
-	Response *http.Response
-
-	// Error, if any.
-	Err error
-
-	// Callback to execute after completion (success or failure).
-	CompletionCallback func(*Task)
-
-	// ProgressCallback is a callback to execute on progress updates
-	ProgressCallback func(int, int)
-
-	// This can be used to cancel the file upload or download if it is no longer needed.
-	Context context.Context
+	// CaptureError saves the error to the task and returns a formatted descriptive error
+	CaptureError(error) error
 }
 
-func (ut *Task) SetProgressCallback(callback func(int, int)) {
-	ut.ProgressCallback = callback
+// TaskCompletionCallback handles the completion callback for a task
+type TaskCompletionCallback struct {
+	CompletionCallback func()
 }
 
-func (ut *Task) SetCompletionCallback(callback func(*Task)) {
-	ut.CompletionCallback = callback
-}
+func (t TaskCompletionCallback) Complete(fts FileTransferStats) {
```

Comment:
I think this doesn't need the `fts` argument, right? I see in `task_default.go` you pass `nil` anyway.

Reply:
yeah it doesn't do anything with the `fts` argument, but without it, it doesn't match the interface. I can instead define a new `Complete` on every `Task` struct to take in the `FileTransferStats` arg and maybe do something with it before calling `TaskCompletionCallback.Complete`- that's probably clearer